src/components/components.js

```diff
@@ -12,28 +12,35 @@ module.exports = function(app) {
         title: 'Layout Components'
       }
     };
-    return {
-      addGroup: function(name, group) {
-        groups[name] = group;
-      },
-      register: function(type, component, group) {
-        if (!components[type]) {
-          components[type] = component;
-        }
-        else {
-          angular.extend(components[type], component);
-        }
 
-        // Set the type for this component.
-        if (!components[type].group) {
-          components[type].group = group || '__component';
-        }
-        components[type].settings.type = type;
-      },
+    function mergeComponents(type, component, group) {
+      if (!components[type]) {
+        components[type] = component;
+      }
+      else {
+        angular.extend(components[type], component);
+      }
+
+      // Set the type for this component.
+      if (!components[type].group) {
+        components[type].group = group || '__component';
+      }
+      components[type].settings.type = type;
+    }
+
+    function addGroup(name, group) {
+      groups[name] = group;
+    }
+
+    return {
+      addGroup: addGroup,
+      register: mergeComponents,
       $get: function() {
         return {
           components: components,
-          groups: groups
+          groups: groups,
+          update: mergeComponents,
```

Comment:
mergeComponets and addGroup are already available during the config phase. Is there a reason you can't use them there instead of during the run phase? We should keep the registration and adding groups to pre-run to make sure the list is complete once we start the run phase.

Typically we do this:
```
angular.module('myapp')
  .config(['formioComponentsProvider', function(){
    formioComponentsProvider.addGroup('test', {title: 'Test'});
    formioComponetnsProvider.register('newfield', definition);
  }])
  .run(['formioComponents', function(formioComponents) {
    formioComponents.groups;
    formioComponents.components;
  }]);
```